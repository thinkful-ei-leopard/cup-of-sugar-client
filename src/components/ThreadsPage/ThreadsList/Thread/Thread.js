import React from 'react';
import { Link } from 'react-router-dom';
import styles from './Thread.module.scss';
import ThreadsApiService from '../../../../services/threads-api-service';
import ThreadsContext from '../../../../contexts/ThreadsContext';
import Confirm from '../../../Confirm/Confirm';
import '@reach/dialog/styles.css';

export default class Thread extends React.Component {
  static contextType = ThreadsContext;

  handleThreadDelete = () => {
    ThreadsApiService.deleteThread(this.props.thread.id);
    this.context.deleteThread(this.props.thread.id);
  };

  findCurrentMessage = () => {
    let messagesForThread = this.context.messages.filter(message => message.thread_id === this.props.thread.id)
    let currentMessage = messagesForThread[0];
    messagesForThread.forEach((message) => {
      if(message.date_modified > currentMessage.date_modified) {
        currentMessage = message
      }
    })
    return currentMessage
  }

  render() {
    const thread = this.props.thread;
    const user = this.props.user;
    let name = '';
    let user_name = '';
    let img_src = '';
    let img_alt = '';
    if (thread.user_id1 === user.id) {
      name = thread.name2;
      user_name = thread.user_name2;
      img_src = thread.img_src2;
      img_alt = thread.img_alt2;
    } else if (thread.user_id1 !== user.id) {
      name = thread.name1;
      user_name = thread.user_name1;
      img_src = thread.img_src1;
      img_alt = thread.img_alt1;
    }

    let currentMessage = this.findCurrentMessage()
    let content = '';
    let date_modifiedMessage = '';
    if(currentMessage) {
      content = currentMessage.content
      if(currentMessage.user_id === user.id)
        date_modifiedMessage = `Sent on: ${currentMessage.date_modified.slice(0, 10)}`
      else if (currentMessage.use_id !== user.id) {
        date_modifiedMessage = `Received on: ${currentMessage.date_modified.slice(0, 10)}`
      }
    }
    
    return (
      <li className={styles.Thread}>
        <div className={styles.threadFlexDiv}>
        <img className={styles.threadAvatarImg} src={img_src} alt={img_alt}></img>
          <div className={styles.threadBlockDiv}>
            <Link to={`/thread/${thread.id}`}>
              <h3 className={styles.threadTitle}>
                {name}({user_name})
              </h3>
            </Link>
            <Confirm title="Confirm" description="Are you sure?">
              {confirm => (
                <button
                  type="delete"
                  className={styles.deleteThreadButton}
                  onClick={confirm(() => {
                    this.handleThreadDelete();
                  })}>
                  <span className={styles.deleteX}>X</span>
                </button>
              )}
            </Confirm>
            <p className={styles.lastMessage}>
              last message: {content}
            </p>
            <p className={styles.lastMessage}>
              {date_modifiedMessage}
            </p>
          </div>
        </div>
      </li>
    );
  }
}
