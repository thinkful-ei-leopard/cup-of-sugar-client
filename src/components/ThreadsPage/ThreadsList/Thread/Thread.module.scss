@import '../../../../styles/_variables';
@import '../../../../styles/_breakpoints';

.Thread {
  //   margin-top: 1rem;
  margin: 0 auto;
  &:hover {
    background-color: var(--light2);
  }
}

.threadTitle {
  margin-bottom: 0.5rem;
  transition: color 0.2s;
  &:hover,
  &:focus {
    color: var(--dark1);
  }
}

.deleteThreadButton {
  position: absolute;
  top: 0.5rem;
  right: 1.5rem;
  font-size: 1rem;
  font-weight: bold;
  color: var(--black);
  transition: color 0.2s;
  background-color: transparent;
  outline: none;
  padding: 0.5rem;
  border: none;
  &:hover {
    color: var(--warning);
    cursor: pointer;
    outline: none;
  }
  &:focus {
    outline: none;
  }
}

.deleteX {
  outline: none;
  background-color: transparent;
}

.lastMessage {
    color: var(--black);
    opacity: .8;
}
