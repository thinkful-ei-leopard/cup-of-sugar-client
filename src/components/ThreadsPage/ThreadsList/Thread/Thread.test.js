import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Thread from './Thread';

it('renders without crashing', () => {
  const div = document.createElement('div');

  ReactDOM.render(
    <Router>
      <Thread thread={{}} user={{}} />
    </Router>,
    div
  );

  ReactDOM.unmountComponentAtNode(div);
});

it('renders the UI as expected', () => {
  const tree = renderer
    .create(
      <Router>
        <Thread thread={{}} user={{}} />
      </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
