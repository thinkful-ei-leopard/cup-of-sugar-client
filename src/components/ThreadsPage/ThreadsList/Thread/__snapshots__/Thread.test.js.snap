// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
<li
  className="Thread"
>
  <a
    href="/thread/undefined"
    onClick={[Function]}
  >
    <div
      className="threadFlexDiv"
    >
      <img
        className="threadAvatarImg"
      />
      <div
        className="threadBlockDiv"
      >
        <h3
          className="threadTitle"
        >
          <span
            className="userRealName"
          >
             
          </span>
          <span
            className="userUserName"
          >
            (
            )
          </span>
        </h3>
        <div
          className="lastMessageContainer"
        >
          <p
            className="lastMessage"
          >
            last message: 
          </p>
          <p
            className="lastMessageContent mobileContent"
          />
          <p
            className="lastMessageContent desktopContent"
          />
        </div>
        <p
          className="lastMessageSent"
        >
          
        </p>
      </div>
      <button
        className="deleteThreadButton"
        onClick={[Function]}
        type="delete"
      >
        <span
          className="deleteX"
        >
          X
        </span>
      </button>
    </div>
  </a>
</li>
`;
